tests: Fix assert for create watchdog unit test

tvyas opened the pull request:
Creating watchdog on a stream handler environment object
without first setting the Systemd WATCHDOG_USEC variable
should be treated normally (not a failure) as the user
might not want to use Systemd or its application watchdog.

Signed-off-by: Tarun Vyas <tarun.vyas@intel.com>

edersondisouza reviewed (commented):
I'm not sure I get what's going on here. Does AVB-SH already supports any other watchdog implementation? If not, how would the only implementation work normally without a mandatory pre-req? If there are more than one implementation, does this test chooses any of them somehow?

tvyas commented:
In the current (and original design), during the init of stream handler object, they try to create the watchdog manager (Systemd based as that's the only one that's supported)  *unconditionally*. The createWatchdog() method then looks for WATCHDOG_USEC. If WATCHDOG_USEC doesn't exist, then it returns a SUCCESS (eIasAvbProcOK) and leaves wdManager as NULL. 
The problem here is that since the createWatchdog() is called unconditionally, it is only supposed to fail when there was an issue with creating the watchdog manager object. If we fail createWatchdog() in case of an absence of WATCHDOG_USEC, then the whole init will fail. Not the nicest thing but that's what the design is doing here. So we sort of have to deal with it.

Hopefully that made sense ;).

edersondisouza commented:
> In the current (and original design), during the init of stream handler object, they try to create the watchdog manager (Systemd based as that's the only one that's supported) _unconditionally_. The createWatchdog() method then looks for WATCHDOG_USEC. If WATCHDOG_USEC doesn't exist, then it returns a SUCCESS (eIasAvbProcOK) and leaves wdManager as NULL.
> The problem here is that since the createWatchdog() is called unconditionally, it is only supposed to fail when there was an issue with creating the watchdog manager object. If we fail createWatchdog() in case of an absence of WATCHDOG_USEC, then the whole init will fail. Not the nicest thing but that's what the design is doing here. So we sort of have to deal with it.
> 
> Hopefully that made sense ;).

Guess I got it now, but that's awful =/
Doesn't it even care about the `cUseWatchdog` thing here? Anyway, I would add a comment explaining what is going on so future us won't be (very) surprised when this bite us back.

tvyas commented:
> > In the current (and original design), during the init of stream handler object, they try to create the watchdog manager (Systemd based as that's the only one that's supported) _unconditionally_. The createWatchdog() method then looks for WATCHDOG_USEC. If WATCHDOG_USEC doesn't exist, then it returns a SUCCESS (eIasAvbProcOK) and leaves wdManager as NULL.
> > The problem here is that since the createWatchdog() is called unconditionally, it is only supposed to fail when there was an issue with creating the watchdog manager object. If we fail createWatchdog() in case of an absence of WATCHDOG_USEC, then the whole init will fail. Not the nicest thing but that's what the design is doing here. So we sort of have to deal with it.
> > Hopefully that made sense ;).
> 
> Guess I got it now, but that's awful =/
Yes it is :|
> Doesn't it even care about the `cUseWatchdog` thing here? Anyway, I would add a comment explaining what is going on so future us won't be (very) surprised when this bite us back.

Not here. This guy just cares about WATCHDOG_USEC. cUseWatchdog kicks in only when individual threads need to register themselves with the watchdog manager, which is being created here.

edersondisouza commented:
One more question: do subsequent calls to watchdog API, which should work when `createdWatchdog` works, still work in this case? Or will people be confused because `createWatchdog` worked but watchdog is not working due missing `WATCHDOG_USEC`?
I'm with the impression that current behaviour was like that because watchdog was not implemented yet, so after it is implemented, behaviour should change to fail on `createWatchdog` when `WATCHDOG_USEC` is not set.

tvyas commented:
> One more question: do subsequent calls to watchdog API, which should work when `createdWatchdog` works, still work in this case? Or will people be confused because `createWatchdog` worked but watchdog is not working due missing `WATCHDOG_USEC`?

The fun part is that there are two separate createWatchdog() calls. One for the StreamHandlerEnvironment object and one for the SystemdWatchdogManager object.

Assuming that WATCHDOG_USEC is set, and one calls StreamHandlerEnvironment->createWatchdog(), the SystemdWatchdogManager object will get populated and then calling a SystemdWatchdogManager ->createWatchdog() will actually add watchdogs to the list.

So, in this case, if one keeps calling StreamHandlerEnvironment->createWatchdog() over and over without setting WATCHDOG_USEC, then they'll keep getting a SUCCESS with the SystemdWatchdogManager still pointing to NULL.


> I'm with the impression that current behaviour was like that because watchdog was not implemented yet, so after it is implemented, behaviour should change to fail on `createWatchdog` when `WATCHDOG_USEC` is not set.

Yea, we could return failed and dump an error message, but then we should make sure that we do not fail the init because of this.

We could also not bother to call StreamHandlerEnvironment->createWatchdog() in the first place if WATCHDOG_USEC is not set. That way, we dont have to worry about init failure.

edersondisouza commented:
I like the "we could return failed and dump an error message, but then we should make sure that we do not fail the init because of this" approach. Is it simple to be done? I'm not sure what "init" means though. "init" is the one who calls `StreamHandlerEnvironment->createWatchdog()`? Like `IasAvbStreamHandler::init()`?

tvyas commented:
Right, it's the `IasAvbStreamHandler::init()` that calls `StreamHandlerEnvironment->createWatchdog()`. Additionally, we need to distinguish that watchdog creation failed b/c of missing WATCHDOG_USEC and not b/c of something else. Then we can continue out _init_ in the former case, and actually fail the _init_ in the latter case.

edersondisouza commented:
This is so obscure - like relating `WATCHDOG_USEC` to having or not the watchdog. That `cUseWatchdog` or so thing is also expected to be set before calling `init`?

tvyas commented:
> That cUseWatchdog or so thing is also expected to be set before calling init?

No, We can have WATCHDOG_USEC and the `StreamHandlerEnvironment->createWatchdog()` call will create the `wdManager` and init will be fine. Without `cUseWatchdog`, none of the threads will actually _add_ themselves to the `wdManager`.
To summarize, we will have the _core_ watchdog enabled but no one will actually be using it.

edersondisouza commented:
If they don't add themselves, them avb-sh will simply be killed after timeout, or nothing will happen? If the second, maybe it's useful some WARN message stating that as `cUseWatchdog` is set but `WATCDHDOG_USEC` not, watchdog won't work.

tvyas commented:
> If they don't add themselves, them avb-sh will simply be killed after timeout, or nothing will happen? If the second, maybe it's useful some WARN message stating that as `cUseWatchdog` is set but `WATCDHDOG_USEC` not, watchdog won't work.

It's the second part. Nothing will happen, b/c the `allInterfacesValid()` will find out that the watchdog list is _empty_, so it'll just return a `true` for the validity part and everything will be warm and cozy.

tvyas commented:
> If the second, maybe it's useful some WARN message stating that as `cUseWatchdog` is set but `WATCDHDOG_USEC` not, watchdog won't work.

I'll include that change with this one then and resubmit.

tvyas committed:
watchdog: Dump WARN if WATCHDOG_USEC is not configured.

Let the user know if cUseWatchdog is set but the environment
variable WATCHDOG_USEC isn't, leading to the watchdog being
unusable.

Signed-off-by: Tarun Vyas <tarun.vyas@intel.com>

tvyas commented:
> > If the second, maybe it's useful some WARN message stating that as `cUseWatchdog` is set but `WATCDHDOG_USEC` not, watchdog won't work.
> 
> I'll include that change with this one then and resubmit.

Pushed a new one. This still isn't the best one, but I wanted to keep it as less invasive as possible and dump sufficient WARNs as well.

edersondisouza commented on private/src/avb_streamhandler/IasAvbStreamHandler.cpp:
```diff
@@ -435,7 +435,21 @@ IasAvbProcessingResult IasAvbStreamHandler::init( const std::string& configName,
         // TODO: Add a config option to receive user input to disable avb_watchdog at runtime.
         //       Otherwise, would not be able to start by executing the demo_app normally
         //       since WATCHDOG_USEC is set by systemctl.
```
This commentary is funny - why not simply do something like `WATCHDOG_USEC=5000 ./avb_streamhandle_demo`? I know it was before your change, but I don't think it makes sense.

tvyas replied:
Yea, I've gone through that comment several times, but left it as is as I wanted to avoid triggering a bigger validation effort.

edersondisouza reviewed (commented):
Looks ok - but in the commit message, "configured through systemd" is misleading, no? I mean, systemd configures because someone added the env var on the unit file. I would simply state something like "env var WATCHDOG_USEC was not configured".
Unless I'm missing something - see my comment on the TODO on code.

tvyas commented:
Looks better now ?
